job/engine.go

```diff
@@ -103,18 +119,46 @@ func (j *Job) Serialize() []byte {
 	return temp
 }
 
-//FIXME: add fault tolerance and security
-func (j *Job) Execute() (interface{}, error) {
+func argsStringified(args []interface{}) string {
+	temp := "("
+	for i, val := range args {
+		if i == len(args)-1 {
+			temp += val.(string) + ""
+		} else {
+			temp += val.(string) + ","
+		}
+	}
+	return temp + ")"
+}
+
+//!TODO: handle retry and retrydelay limit
+func (j *Job) Execute(exec *JobExec) {
+	exec.SetStatus(RUNNING)
+	r := exec.GetRetries()
+retry:
 	env := vm.NewEnv()
 	start := time.Now()
-	result, err := env.Execute(string(j.GetSource()))
-	exec := JobExec{
-		Timestamp: time.Now().Unix(),
-		Duration:  time.Now().Sub(start).Nanoseconds(),
-		Err:       err,
-		Result:    result,
-		By:        []byte("0000"), //FIXME: replace with real ID
+	var result interface{}
+	var err error
+	if len(exec.GetArgs()) == 0 {
+		result, err = env.Execute(string(helpers.Decode64(j.GetTask())) + "\n" + j.GetName() + "()")
+	} else {
+		result, err = env.Execute(string(helpers.Decode64(j.GetTask())) + "\n" + j.GetName() + argsStringified(exec.GetArgs()))
+	}
+
+	if r != 0 && err != nil {
+		r--
+		time.Sleep(exec.GetRetryDelay() * time.Second)
+		exec.SetStatus(RETRYING)
+		glg.Info("Retrying job - " + j.GetID())
+		goto retry
 	}
-	j.AddExec(exec)
-	return result, err
+	exec.SetTimestamp(time.Now().Unix())
+	exec.SetDuration(time.Duration(time.Now().Sub(start).Nanoseconds()))
+	exec.SetErr(err)
+	exec.SetResult(result)
+	exec.setHash()
+	exec.SetStatus(FINISHED)
+	j.AddExec(*exec)
+	// return &exec
```

Comment:
don't you want to return an error here ?

Reply:
The error would be handled by the user. It shouldn't be our problem.

Reply:
how does the user know an error has occurred ?

Reply:
Just like an API call, you know about it when it returns

Reply:
Okay we might still need some form of error handling if things ever fail on our side....its fine in \this case

Reply:
Yeah for sure